runtime/testy/runtime.py

```diff
@@ -91,17 +91,18 @@ def nonTestModePrint(*args):
           controlState = "auto"
           continue
         elif newBadThing.event == BAD_EVENTS.ENTER_IDLE and controlState != "idle":
+          controlState = "idle"
           break
         print(newBadThing.event)
         nonTestModePrint(newBadThing.data)
         if newBadThing.event in restartEvents:
+          restartCount += 1
           if (not emergency_stopped and newBadThing.event is BAD_EVENTS.EMERGENCY_STOP):
             emergency_stopped = True #somehow kill student code using other method? right now just restarting on e-stop
           break
       stateQueue.put([SM_COMMANDS.RESET, []])
       terminate_process(PROCESS_NAMES.STUDENT_CODE)
       controlState = "idle"
```

Comment:
It looks like you're moving all control state changes into the while loop, so I think you should move this one too (and maybe add an assert if the controlstate must always be idle at this point).